### ISSUE TITLE
General Help: How can I merge commits into my forked repo?

### ISSUE THREAD

**Issue by ghost** [domoticz/domoticz#3744]
I forked the domoticz repo a while ago so I could create a PR, but now it's 173 commits behind.

Can someone just give me a helpful direction as to how I can 'sync' my forked repo so I can keep up to date with the latest commits to the domoticz repo?

Thanks in advance

**Comment by petergebruers** [comment]
I can try... You have any experience with either "merge" or "rebase"? Ever solved a "conflict" before?

**Comment by RDols** [comment]
There are several ways to do. This is one of the command-line options.

In the case you don't have a local clone of your repository:
`md domoticz`
`git clone https://github.com/leecollings/domoticz.git domoticz`
`cd domoticz`
`git checkout -B origin/development`

Add domoticz source to your local repository, only need to do this once:
`git remote add domoticz https://github.com/domoticz/domoticz.git`

Next steps you have to do every time you want an update:
`git fetch --all`
`git merge domoticz/development`
`git push --set-upstream origin origin/development`

**Comment by petergebruers** [comment]
@leecollings try what @RDols said... But maybe, if you are inexperienced with git, you want more guidance? Please understand I do not want to underestimate your skills... I am unexperienced myself but that might be an advantage because I stumble on similar issues.

I'm not a Domoticz developer and not a seasoned Git users either... but I use Git frequently so I might be able to help.

First of all, it might NOT matter at all, if open a PR based on your local branch... Github will tell you if it is able to merge automatically or not. That depends on how many parts you've edited and have been touched by others. Which is quite possible if you forked 173 commits ago... Maybe that is the reason why you are asking "how to sync"?

If github tells you it has a merge conflict, you can fix that and force push (if you do a rebase as I'll explain below) to that branch. On the other hand, if you haven't opened a PR yet and expect issues, I would fix them locally before pushing.

If you are an infrequent user, a GUI might come in handy.

I've tried a few git Gui's over the years and I've settled on:

https://www.syntevo.com/smartgit/

You might appreciate such a tool if you need to solve conflicts because of its "merge editor" and easier ways to select commits and compare.

This is not a buying advice. But it is free for non commercial use :wink:

First of all... You have your work in a separate branch of your local repository, right? And your work was committed to this branch? Good!

In that case I would start by making a copy of that branch. It is not strictly necessary but if you are new to all this, this step will make you feel safer. Call it "test". Menu branch, add branch.

Now check out "Development" on your repo, which is "tracking" the branch "Development" of your github repository, then one that is a fork of the (canonical) Domoticz repository

If you are a command-line-guy... These commands are basically behind the scenes of all the menu commands in SmartGit.

This command will help you find the correct branch, take note of the "double vv":

`git branch -vv`

```
  check_already_added               3d21a9dc1 OZW, Update detection of duplicate/invalid DeviceID
  conn-timeout                      cc1098f01 WIP
  development                       ac3d2290d [domoticz-pgbr/development] Merge branch 'development' of github.com:domoticz/domoticz into development
* development-domoticz-public       95e2be251 [domoticz-public/development] Jsoncpp deprecate fix (#3741)
```

If you want to double check the remotes, the info between square brackets is here:

`git remote -v`

```
domoticz-pgbr	https://github.com/petergebruers/domoticz.git (fetch)
domoticz-pgbr	https://github.com/petergebruers/domoticz.git (push)
domoticz-private	https://github.com/petergebruers/domoticz-private.git (fetch)
domoticz-private	https://github.com/petergebruers/domoticz-private.git (push)
domoticz-public	https://github.com/domoticz/domoticz.git (fetch)
domoticz-public	https://github.com/domoticz/domoticz.git (push)
```

So In my case the branch I would like to keep up-to-date is called "development-domoticz-public"

I am already on that branch but if I wasn't then I would do:

`git checkout development-domoticz-public`

Then to merge info from Domoticz:

`git pull`

```
remote: Enumerating objects: 12, done.
remote: Counting objects: 100% (12/12), done.
()...)

 2 files changed, 2 insertions(+), 2 deletions(-)
Current branch development-domoticz-public is up to date.
```

Now I have the latest commit and I am up-to-date on this branch.

Now suppose I have been working on branch "conn-timeout" and it got behind by several commits, and I am working on this all by myself, then I would "rebase" my work on top of the newest commit. Rebase has the advantage over merge that it really changes commits so they appear as if you started working on this from the latest version. It is magic. But "changes commits" is a non-go if you are collaborating on this with somebody else via "git". It is possible though to rebase and force push to github, some public repositories even want you to do that. The discussion of "rebase" versus "merge" is long.

Let's rebase my "conn-timeout" branch. And as I've said, if you are new, you'll feel more comfortable if you first make a copy of your work

`git checkout -b new_branch old_branch`

In my case

`git checkout -b test conn-timeout`

Git says "Switched to a new branch 'test'

Now it is time to rebase... I'll tell you I do this in "SmartGit" if I expect my work to be "in conflict" with the work that has been done on the public Domoticz branch.

`git rebase development-domoticz-public`

```
M	webserver/connection.cpp
Falling back to patching base and 3-way merge...
Auto-merging webserver/connection.cpp
CONFLICT (content): Merge conflict in webserver/connection.cpp
Auto-merging hardware/ASyncTCP.cpp
error: Failed to merge in the changes.
Patch failed at 0001 WIP
hint: Use 'git am --show-current-patch' to see the failed patch
Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
(...)
```

I took this particular example because I was 99% sure, my branch would conflict, because I knew another user was working on this.

This is when a nice GUI can help, although you can see/do everything in test mode.

The part between the "<<<<<<<" and ">>>>>>>" tells you that a part get edited and those edits cannot be reconciliated.

```txt
<<<<<<< HEAD
				if (connection_type==connection_websocket)
				{
					websocket_parser.Stop();
				}
				connection_manager_.stop(shared_from_this());
=======
				socket_->shutdown(boost::asio::ip::tcp::socket::shutdown_both);
>>>>>>> WIP
```

Sometimes you are lucky and those conflicts are small and you can easilly select one of both sources.

Sometimes... both sides have diverged so much that you'll spend a lot of time fixing them.

But you're on the "test" branch and can edit whatever you want, build and test until it is OK again.

If all conflict markers have been removed and everything is committed, you can finish the rebase:

`git rebase --continue`

Then you can push that branch (the fact that it is called test does not matter). But if you repeat this process after the PR you'll have to "force push". If other users work on that branch, "force push" will make them unhappy. If this is your PR and you work alone.

Also it is worth mentioning that if your history contains a lot of commits for a relatively simple PR, it might make sense to do a "squash" of the commits. But imho that is not important right now. The owners of the repo can squash commits if they want to. It is a topic on its own.

Keep your original branch (in my case "conn-timeout") for a while, because if something turns out to be different than expected... You can checkout that branch, build it and test and debug your program... Then compare with "test" branch.

Please tell me if this helps. Or if it is good or bad advice.
